### DISCUSSION TITLE
WDIO Config file

### DISCUSSION THREAD

**Discussion by bajipogadadanda** [webdriverio/codemod#102] [General]
Hi,
I need to define custom properties in wdio.confg file(ex:apmurl along with BaseURL), I  need to get it in the Test case definition.
would that be possible?
here is the actual requirement.
1.I need to define JS object, also other properties as below and need to get it for various spec files.
exports.config = {
testURL:'https://webdriver.io/',
params: {
       A: 'OFF',
       B: 2000,
       C:'testDataEnv'
      },
}

I read that it was possible prior WDIO 8 version.
Pls let me know how can it be attained for latest version also.

**Comment by christian-bromann** [comment]
It is not recommended to store arbitrary data in the WebdriverIO config. Use a different file for this.

**Reply by bajipogadadanda** [reply]
Ok Thx Chris!
 is there anyway to define global values in wdio, or any hook to initialize the global values before execution starts in config file? 
My need is that, few variables/properties to be initialized before WDIO executions starts and it shd be available for all the Spec files/any spec file in the Fraemwork.

**Reply by christian-bromann** [reply]
use the `before` hook at it gets executed within the worker process.

**Reply by bajipogadadanda** [reply]
Hope you are referring below at wdio.config.js file. 
I tried below and tried access this variable in spec file. But, its not working, getting reference error "referenceError: testEnvReferenceAPI is not defined".
I think it a procedural issue from my end. Can you refer me if any example for it.
Also, would onprepare also helps for it?

before: function (capabilities, specs) {
                   var testEnvReferenceAPI="test/psotAPI"
    },

**Reply by christian-bromann** [reply]
> var testEnvReferenceAPI="test/psotAPI"

This does not define a global variable

**Reply by bajipogadadanda** [reply]
I thought so, can you pls refer me an example how to define few global variables/properties(ex: reference API, auth URLs and e.t.c) at wdio.config.js file level under before, or onPrepare(if its appropriate) and access them through a spec file.
It helps me a lot!!

**Reply by christian-bromann** [reply]
Have you tried to Google this? These are fundamental that can be found in tutorials on the internet.

**Comment by bajipogadadanda** [comment]
Yes Chris. I did.
I am using mocha with JavaScript.
so, there are many answers, whereas I tried using 'global'(ex: global.EnvAPI) , also few more . Unfortunately, non worked(same reference error). There are few examples , but they meant for older versions(I am using V8).

Config file:

before: function (capabilities, specs) {
             global.REFERENCEAPI="test/psotAPI"
          
    },

Spec file:
describe('Test spec', async function () {
       it('tests something here, async function() {
       console.log("Test Global Var:", REFERENCEAPI)
    
   });
});

**Reply by bajipogadadanda** [reply]
Hi Chris,
I have tried various other options as well, example as below.
None of them worked, so I would like to know if this feature available for WebdriberIO 8 or not, or shd we use other hooks like onPrepare.
I m using WDIO8 with Mocha, and JS
Note: I learned that params Object, global object also is being supported for Ver8 WDIOconfig file.
Result: no error, but value has not printed (printed as undefined), i.e its not accessing the assigned value.

Config File:

before: function (capabilities, specs) {
        
        global.API='test'
}

Spec File:

describe('Test spec', async function () {
 // before(() => {
    //     api = API
    //     api = global.API
    // })
const api=global.API
it('tests something here, async function() {
console.log("Test Global Var:", api)

});
});

**Reply by christian-bromann** [reply]
Instead of using global variables which are kind of a bad pattern you can also use the shared store service: https://webdriver.io/docs/shared-store-service/

**Reply by bajipogadadanda** [reply]
Thx Chris!
I never used this service and not much info on internet, so its a learning for me :) !!
I understand it in theory, I have tried a sample as below, these changes have not impacted my existing specs and they are good. However, I couldn’t get the clue on how these values can be accessed for the Spec files?  Thought check if you can provide a light on it.

//WDIO Config file:

const shared = require('@wdio/shared-store-service')

exports.config = {
    onPrepare: [async function (config, capabilities) {
        await setValue('testKey', 'testValue')
    }],
  
// after: async () => {
    //     const value = await getValue('testKey')},
    

   beforeSession: async (conf) => {
        conf.value = await getValue('testKey');
    },


}

services: [
        chromedriver,
       ['shared-store',{}]  
      
    ],

**Reply by christian-bromann** [reply]
I think the documentation is clear on this: https://webdriver.io/docs/shared-store-service/#get-values
